lib/safira_web/live/sponsor/scanner/index.ex

```diff
@@ -0,0 +1,166 @@
+    {:ok,
+     socket
+     |> assign(:current_page, :scanner)
+     |> assign(:modal_data, nil)
+     |> assign(:given_list, [])}
+  end
+
+  @impl true
+  def handle_params(_params, _url, socket) do
+    badge_id = socket.assigns.current_user.company.badge_id
+
+    if is_nil(badge_id) do
+      {:noreply,
+       socket
+       |> put_flash(:error, "Company does not have badge")
+       |> push_navigate(to: ~p"/sponsor")}
+    else
+      badge = Contest.get_badge!(badge_id)
+      {:noreply, socket |> assign(:badge, badge)}
+    end
+  end
+
+  @impl true
+  def handle_event("scan", data, socket) do
+    case safely_extract_id_from_url(data) do
+      {:ok, id} -> process_scan(id, socket)
+      {:error, _} -> {:noreply, assign(socket, :modal_data, :invalid)}
+    end
+  end
+
+  @impl true
+  def handle_event("close-modal", _, socket) do
+    {:noreply, socket |> assign(:modal_data, nil)}
+  end
+
+  defp process_scan(id, socket) do
+    if id in socket.assigns.given_list do
+      {:noreply, socket}
+    else
+      check_credential(id, socket)
+    end
+  end
+
+  defp check_credential(id, socket) do
+    if Accounts.credential_exists?(id) do
+      handle_attendee_lookup(id, socket)
+    else
+      {:noreply, assign(socket, :modal_data, :not_found)}
+    end
+  end
+
+  defp handle_attendee_lookup(id, socket) do
+    case Accounts.get_attendee_from_credential(id) do
+      nil ->
+        {:noreply, assign(socket, :modal_data, :not_linked)}
+
+      attendee ->
+        handle_redeem_badge(
+          %{
+            badge_id: socket.assigns.badge.id,
+            attendee: attendee,
+            credential_id: id
+          },
+          socket
+        )
+    end
+  end
+
+  defp handle_redeem_badge(
+         %{badge_id: badge_id, attendee: attendee, credential_id: credential_id},
+         socket
+       ) do
+    badge = Contest.get_badge!(badge_id)
+
+    case Contest.redeem_badge(badge, attendee, nil) do
+      {:ok, _} ->
+        {:noreply,
+         socket
+         |> assign(:modal_data, nil)
+         |> assign(:given_list, [credential_id | socket.assigns.given_list])}
+
+      {:error, "attendee already has this badge"} ->
+        {:noreply,
+         socket
+         |> assign(:modal_data, :already_redeemed)
+         |> assign(:given_list, [credential_id | socket.assigns.given_list])}
+
+      {:error, _} ->
+        {:noreply, socket |> assign(:modal_data, :not_linked)}
+    end
+  end
+
+  defp error_message(:already_redeemed),
+    do: gettext("Attendee already has this badge! (400)")
+
+  defp error_message(:not_found),
+    do: gettext("This credential is not registered in the event's system! (404)")
+
+  defp error_message(:not_linked),
+    do: gettext("This credential is not linked to any attendee! (400)")
+
+  defp error_message(:invalid), do: gettext("Not a valid credential! (400)")
+
+  defp error_message(:out_of_time),
+    do: gettext("This badge is not currently available for redemption! (400)")
```

Comment:
```suggestion
  defp error_message(:out_of_time),
    do: gettext("This badge is not currently available for redemption! (400)")
```

Can't happen in this page.